nx/lib/nx/binary_backend.ex

```diff
@@ -1038,6 +1038,138 @@ defmodule Nx.BinaryBackend do
     from_binary(out, output_data)
   end
 
+  @impl true
+  def qr(%{shape: {_m, k} = shape_q} = q, %{shape: {k, n}} = r, opts) do
+    mode = opts[:mode]
+
+    keep_reflectors = mode == :raw
+    calculate_q = mode in [:reduced, :complete, :raw]
+
+    q =
+      if calculate_q do
+        identity(q, shape: shape_q)
+      end
+
+    {q, r, reflectors} =
+      for i <- 0..(n - 1), reduce: {q, r, []} do
+        {q, r, reflectors} ->
+          %{shape: {reflector_n, reflector_m}} =
+            reflector = householder_reflector(r[[i..(k - 1), i]])
+
+          shape_h = {k, k}
+          h = identity(as_type(reflector, r), shape: shape_h)
+
+          h_bin =
+            for col <- 0..(reflector_m - 1), row <- 0..(reflector_n - 1), reduce: to_binary(h) do
+              acc ->
+                set_value_at_index(
+                  acc,
+                  h.type,
+                  shape_h,
+                  [row + i, col + i],
+                  reflector[[row, col]]
+                )
+            end
+
+          h = from_binary(h, h_bin)
+
+          reflectors =
+            if keep_reflectors do
+              [h | reflectors]
+            else
+              reflectors
+            end
+
+          q =
+            if calculate_q do
+              Nx.dot(q, h)
+            end
+
+          r = Nx.dot(h, r)
+
+          {q, r, reflectors}
+      end
+
+    case mode do
+      :complete ->
+        {q, r}
+
+      :reduced ->
+        {q, r[[0..(k - 1), 0..(n - 1)]]}
+
+      :r ->
+        r
+
+      :raw ->
+        {q, r, reflectors}
+    end
+  end
+
+  defp householder_reflector(%{shape: {n}} = a) do
+    a_0 = a[0]
+
+    v =
+      %{type: {_, num_bits} = type_v} =
+      if Nx.sign(a_0) == -1 do
+        Nx.divide(a, Nx.subtract(a_0, Nx.norm(a)))
+      else
+        Nx.divide(a, Nx.add(a_0, Nx.norm(a)))
```

Comment:
Just a note for this stuff, we try to avoid calls to the `Nx` namespace, so you should either try to work on the raw binaries OR you can use the implementations in this module. It will just save you the overhead of dispatching here.